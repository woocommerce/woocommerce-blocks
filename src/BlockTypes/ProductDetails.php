<?php
namespace Automattic\WooCommerce\Blocks\BlockTypes;

use Automattic\WooCommerce\Blocks\Utils\StyleAttributesUtils;

/**
 * ProductDetails class.
 */
class ProductDetails extends AbstractBlock {
	/**
	 * Block name.
	 *
	 * @var string
	 */
	protected $block_name = 'product-details';

	/**
	 * It isn't necessary register block assets because it is a server side block.
	 */
	protected function register_block_type_assets() {
		return null;
	}

	/**
	 * Render the block.
	 *
	 * @param array    $attributes Block attributes.
	 * @param string   $content Block content.
	 * @param WP_Block $block Block instance.
	 *
	 * @return string Rendered block output.
	 */
	protected function render( $attributes, $content, $block ) {
		$tabs = $this->render_tabs();

		$classname = $attributes['className'] ?? '';

		$classes_and_styles = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );

		return sprintf(
			'<div class="wp-block-woocommerce-product-details %1$s %2$s">
				<div style="%3$s">
					%4$s
				</div>
			</div>',
			esc_attr( $classes_and_styles['classes'] ),
			esc_attr( $classname ),
			esc_attr( $classes_and_styles['styles'] ),
			$tabs
		);
	}

	/**
	 * Gets the tabs with their content to be rendered by the block.
	 *
	 * @return string The tabs html to be rendered by the block
	 */
	protected function render_tabs() {
		ob_start();
		rewind_posts();
		while ( have_posts() ) {
			the_post();
			woocommerce_output_product_data_tabs();
		}

		$tabs = ob_get_clean();

		return $tabs;
	}
}
