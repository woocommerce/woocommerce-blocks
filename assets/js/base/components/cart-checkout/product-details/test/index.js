/**
 * External dependencies
 */
import TestRenderer from 'react-test-renderer';

/**
 * Internal dependencies
 */
import ProductDetails from '..';

describe( 'ProductDetails', () => {
	test( 'should render details', () => {
		const details = [
			{ name: 'Lorem', value: 'Ipsum' },
			{ name: 'LOREM', value: 'Ipsum', display: 'IPSUM' },
			{ value: 'Ipsum' },
		];
		const component = TestRenderer.create(
			<ProductDetails details={ details } />
		);

		expect( component.toJSON() ).toMatchSnapshot();
	} );

	test( 'should not render hidden details', () => {
		const details = [
			{ name: 'Lorem', value: 'Ipsum', hidden: true },
			{ name: 'LOREM', value: 'Ipsum', display: 'IPSUM' },
		];
		const component = TestRenderer.create(
			<ProductDetails details={ details } />
		);

		expect( component.toJSON() ).toMatchSnapshot();
	} );

	test( 'should not rendering anything if all details are hidden', () => {
		const details = [
			{ name: 'Lorem', value: 'Ipsum', hidden: true },
			{ name: 'LOREM', value: 'Ipsum', display: 'IPSUM', hidden: true },
		];
		const component = TestRenderer.create(
			<ProductDetails details={ details } />
		);

		expect( component.toJSON() ).toMatchSnapshot();
	} );

	test( 'should not rendering anything if details is an empty array', () => {
		const details = [];
		const component = TestRenderer.create(
			<ProductDetails details={ details } />
		);

		expect( component.toJSON() ).toMatchSnapshot();
	} );
} );
