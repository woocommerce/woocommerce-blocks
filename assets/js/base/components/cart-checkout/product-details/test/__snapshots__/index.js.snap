// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails should not render hidden details 1`] = `
<ul
  className="wc-block-components-product-details"
>
  <li
    className="wc-block-components-product-details__item"
  >
    <span
      className="wc-block-components-product-details__name"
    >
      LOREM
      :
    </span>
     
    <span
      className="wc-block-components-product-details__value"
    >
      IPSUM
    </span>
  </li>
</ul>
`;

exports[`ProductDetails should not rendering anything if all details are hidden 1`] = `null`;

exports[`ProductDetails should not rendering anything if details is an empty array 1`] = `null`;

exports[`ProductDetails should render details 1`] = `
<ul
  className="wc-block-components-product-details"
>
  <li
    className="wc-block-components-product-details__item"
  >
    <span
      className="wc-block-components-product-details__name"
    >
      Lorem
      :
    </span>
     
    <span
      className="wc-block-components-product-details__value"
    >
      Ipsum
    </span>
  </li>
  <li
    className="wc-block-components-product-details__item"
  >
    <span
      className="wc-block-components-product-details__name"
    >
      LOREM
      :
    </span>
     
    <span
      className="wc-block-components-product-details__value"
    >
      IPSUM
    </span>
  </li>
  <li
    className="wc-block-components-product-details__item"
  >
    
    <span
      className="wc-block-components-product-details__value"
    >
      Ipsum
    </span>
  </li>
</ul>
`;
